airbyte-cdk/python/airbyte_cdk/sources/async_source.py

```diff
@@ -0,0 +1,103 @@
+import collections
+import itertools
+import time
+from abc import ABC, abstractmethod
+from typing import Iterator, Union, Mapping
+
+import pendulum
+from airbyte_cdk.sources import AbstractSource
+from pydantic.tools import Any
+
+
+class JobWaitTimeout(Exception):
+    pass
+
+
+class AsyncJob(ABC):
+    """Total time allowed for job to run, in case it is None the job will run endlessly"""
+    job_wait_timeout = None
+
+    @property
+    @abstractmethod
+    def job_sleep_interval(self) -> pendulum.Duration:
+        """Sleep interval after each check of job status"""
+        return pendulum.duration(seconds=5)
+
+    @abstractmethod
+    def start_job(self) -> None:
+        """Create async job and return"""
+
+    @abstractmethod
+    def completed_successfully(self) -> bool:
+        """Something that will tell if job was successful"""
+
+    def should_retry(self, exc: Exception) -> bool:
+        """Tell if the job should be restarted when the following exception caught"""
+        return False
+
+    def _wait_for_job(self):
+        """Actual waiting for job to finish"""
+        start_time = pendulum.now()
+
+        while self.job_wait_timeout and pendulum.now() - start_time > self.job_wait_timeout:
+            if self.completed_successfully():
+                return
+            time.sleep(self.job_sleep_interval.in_seconds())
+
+        raise JobWaitTimeout("Waiting for job more than allowed")
+
+    def fetch_result(self) -> Any:
```

Comment:
Maybe this is better named as `fetch_as_completed` or similar? Since we're not actually getting the data.

Reply:
as discussed replaced it with wait_completion